### PR TITLE
fix: set optional validate for field when update/add  user/staff

### LINKED ISSUE #723: Editor, name, ID number, address should not be required
编辑员工，名称，身份证号，地址，职位应为非必填项
image

### PR DESCRIPTION
**Pull request by xyfy**
fix: set optional validate for field when update/add  user/staff

1. for user, the idcard、postion、companyName、department  is optional
2. for staff, the idcard、position、name、address is optional
3. update git ignore
4. Service use the same validator with fronted

resolves  #723  #721 #718 #717

### CODE REVIEW

File: src/Services/Masa.Auth.Service.Admin/Application/Subjects/Commands/UpdateStaffCommandValidator.cs [hunk 1]
```diff
@@ -7,13 +7,6 @@ public class UpdateStaffCommandValidator : AbstractValidator<UpdateStaffCommand>
...
     {
-        RuleFor(command => command.Staff.JobNumber).Required().MaximumLength(20);
-        RuleFor(command => command.Staff.DisplayName).Required().MaximumLength(50);
-        RuleFor(command => command.Staff.Name).ChineseLetter().MaximumLength(20);
-        RuleFor(command => command.Staff.PhoneNumber).Required().Phone();
-        RuleFor(command => command.Staff.Email).Email();
-        RuleFor(command => command.Staff.IdCard).IdCard();
-        RuleFor(command => command.Staff.CompanyName).ChineseLetter().MaximumLength(50);
-        RuleFor(command => command.Staff.Position).ChineseLetterNumber().MaximumLength(20);
+        RuleFor(command => command.Staff).SetValidator(new UpdateStaffValidator());
```
Comment (wuweilaiya) [inline src/Services/Masa.Auth.Service.Admin/Application/Subjects/Commands/UpdateStaffCommandValidator.cs:10]: For example, a nickname is not required on the server side,please refer to the source code to find out the difference,fix it.

File: src/Services/Masa.Auth.Service.Admin/Application/Subjects/Commands/AddUserCommandValidator.cs [hunk 2]
```diff
@@ -7,14 +7,6 @@ public class AddUserCommandValidator : AbstractValidator<AddUserCommand>
...
-        RuleFor(command => command.User.DisplayName).MaximumLength(50);
-        RuleFor(command => command.User.Name).ChineseLetterNumber().MaximumLength(20);
-        RuleFor(command => command.User.PhoneNumber).Phone();
-        RuleFor(command => command.User.Email).Email();
-        RuleFor(command => command.User.IdCard).IdCard();
-        RuleFor(command => command.User.CompanyName).ChineseLetterNumber().MaximumLength(50);
-        RuleFor(command => command.User.Position).ChineseLetterNumber().MaximumLength(20);
-        RuleFor(command => command.User.Account).MaximumLength(50);
-        RuleFor(command => command.User.Password).AuthPassword();
+        RuleFor(command => command.User).SetValidator(new AddUserValidator());
```
Comment (wuweilaiya) [inline src/Services/Masa.Auth.Service.Admin/Application/Subjects/Commands/AddUserCommandValidator.cs:10]: Form validation rules for client and server are different,for example, a nickname is not required on the server side.

File: src/Services/Masa.Auth.Service.Admin/Application/Subjects/Commands/UpdateUserCommandValidator.cs [hunk 3]
```diff
@@ -7,12 +7,6 @@ public class UpdateUserCommandValidator : AbstractValidator<UpdateUserCommand>
...
     public UpdateUserCommandValidator()
     {
-        RuleFor(command => command.User.DisplayName).Required().MaximumLength(50);
-        RuleFor(command => command.User.Name).ChineseLetterNumber().MaximumLength(20);
-        RuleFor(command => command.User.PhoneNumber).Phone();
-        RuleFor(command => command.User.Email).Email();
-        RuleFor(command => command.User.IdCard).IdCard();
-        RuleFor(command => command.User.CompanyName).ChineseLetterNumber().MaximumLength(50);
-        RuleFor(command => command.User.Position).ChineseLetterNumber().MaximumLength(20);
+        RuleFor(command => command.User).SetValidator(new UpdateUserValidator());
```
Comment (wuweilaiya) [inline src/Services/Masa.Auth.Service.Admin/Application/Subjects/Commands/UpdateUserCommandValidator.cs:10]: Form validation rules for client and server are different,for example, a avatar is not required on the server side.

File: src/Services/Masa.Auth.Service.Admin/Application/Subjects/Commands/AddStaffCommandValidator.cs [hunk 4]
```diff
@@ -7,14 +7,6 @@ public class AddStaffCommandValidator : AbstractValidator<AddStaffCommand>
...
-        RuleFor(command => command.Staff.JobNumber).Required().MaximumLength(20);
-        RuleFor(command => command.Staff.DisplayName).MaximumLength(50);
-        RuleFor(command => command.Staff.Name).ChineseLetter().MaximumLength(20);
-        RuleFor(command => command.Staff.PhoneNumber).Required().Phone();
-        RuleFor(command => command.Staff.Email).Email();
-        RuleFor(command => command.Staff.IdCard).IdCard();
-        RuleFor(command => command.Staff.CompanyName).ChineseLetter().MaximumLength(50);
-        RuleFor(command => command.Staff.Position).ChineseLetterNumber().MaximumLength(20);
-        RuleFor(command => command.Staff.Password).AuthPassword();
+        RuleFor(command => command.Staff).SetValidator(new AddStaffValidator());
```
Comment (wuweilaiya) [inline src/Services/Masa.Auth.Service.Admin/Application/Subjects/Commands/AddStaffCommandValidator.cs:10]: For example, a nickname is not required on the server side,for example:avatar、password,please refer to the source code to find out the difference,fix it.

File: src/Contracts/Masa.Auth.Contracts.Admin/Subjects/Validator/UpdateStaffValidator.cs [hunk 5]
```diff
@@ -11,13 +11,14 @@ public UpdateStaffValidator()
...
             .NotEmpty().WithMessage("NickName is required")
             .ChineseLetterNumber().MinimumLength(2).MaximumLength(50).OverridePropertyName("NickName");
-        RuleFor(staff => staff.Name).ChineseLetterNumber().MinimumLength(2).MaximumLength(50);
+        RuleFor(staff => staff.Name).ChineseLetterNumber().MinimumLength(2).MaximumLength(50)
+            .When(u => !string.IsNullOrEmpty(u.Name));
         RuleFor(staff => staff.PhoneNumber).Required().Phone();
         RuleFor(staff => staff.Email).Email();
-        RuleFor(staff => staff.IdCard).IdCard();
-        RuleFor(staff => staff.Address.Address).MinimumLength(8).MaximumLength(100);
+        RuleFor(staff => staff.IdCard).IdCard().WithMessage("Id Card format is incorrect").When(u => !string.IsNullOrEmpty(u.IdCard));
```
Comment (wuweilaiya) [inline src/Contracts/Masa.Auth.Contracts.Admin/Subjects/Validator/UpdateStaffValidator.cs:18]: 校验身份证和校验是否必填应该是分开的，到处使用When来填补这个问题是否合理？

**Review by wuweilaiya — CHANGES_REQUESTED** [review]

File: src/Contracts/Masa.Auth.Contracts.Admin/Subjects/Validator/UpdateStaffValidator.cs (hunk 5 above)
Comment (xyfy) [reply src/Contracts/Masa.Auth.Contracts.Admin/Subjects/Validator/UpdateStaffValidator.cs:18]: 没有找到一种可以标记可选的办法，理论上应该是Idcard在null或者空的情况下返回pass，靠notempty来控制是否可空，但是目前Idcard没有支持

Comment (wuweilaiya) [reply src/Contracts/Masa.Auth.Contracts.Admin/Subjects/Validator/UpdateStaffValidator.cs:18]: 那可以自定义封装解决这个问题，比如把你的when not empty封装下或者重写校验器

File: src/Contracts/Masa.Auth.Contracts.Admin/Subjects/Validator/AddStaffValidator.cs [hunk 6]
```diff
@@ -11,13 +11,13 @@ public AddStaffValidator()
         RuleFor(staff => staff.DisplayName)
             .NotEmpty().WithMessage("NickName is required")
             .ChineseLetterNumber().MinimumLength(2).MaximumLength(50).OverridePropertyName("NickName");
-        RuleFor(staff => staff.Name).ChineseLetterNumber().MinimumLength(2).MaximumLength(50);
+        When(staff => !string.IsNullOrEmpty(staff.Name), () => RuleFor(staff => staff.Name).ChineseLetterNumber().MinimumLength(2).MaximumLength(50));
```
Comment (wuweilaiya) [inline src/Contracts/Masa.Auth.Contracts.Admin/Subjects/Validator/AddStaffValidator.cs:14]: 请提供类似WhenNotEmpty()方法，此方法会自动识别用户写的Expression，无需手动判断staff => !string.IsNullOrEmpty(staff.Name)

**Review by wuweilaiya — CHANGES_REQUESTED** [review]

**Review by wuweilaiya — APPROVED** [review]

**Pull request merged**